creusot-contracts/src/std/option.rs

```diff
@@ -35,63 +35,276 @@ extern_spec! {
-                fn unwrap(self) -> T;
+                fn expect(self, msg: &str) -> T;
 
                 #[pure]
                 #[requires(self != None)]
                 #[ensures(Some(result) == self)]
-                fn expect(self, msg: &str) -> T;
+                fn unwrap(self) -> T;
 
                 #[pure]
                 #[ensures(self == None ==> result == default)]
                 #[ensures(self == None || self == Some(result))]
                 fn unwrap_or(self, default: T) -> T;
 
-                #[pure]
-                #[ensures(*self == None ==> result == None && ^self == None)]
-                #[ensures(
-                    *self == None
-                    || exists<r: &mut T> result == Some(r) && *self == Some(*r) && ^self == Some(^r)
-                )]
-                fn as_mut(&mut self) -> Option<&mut T>;
+                #[requires(self == None ==> f.precondition(()))]
+                #[ensures(match self {
+                    None => f.postcondition_once((), result),
+                    Some(t) => result == t
+                })]
+                fn unwrap_or_else<F>(self, f: F) -> T
+                where
+                    F: FnOnce() -> T;
 
-                #[pure]
-                #[ensures(*self == None ==> result == None)]
-                #[ensures(
-                    *self == None || exists<r: &T> result == Some(r) && *self == Some(*r)
-                )]
-                fn as_ref(&self) -> Option<&T>;
+                #[ensures(self == None ==> result.is_default())]
+                #[ensures(self == None || self == Some(result))]
+                fn unwrap_or_default(self) -> T
+                where
+                    T: Default;
+
+                #[requires(match self {
+                    None => true,
+                    Some(t) => f.precondition((t,)),
+                })]
+                #[ensures(match self {
+                    None => result == None,
+                    Some(t) => exists<r: _> result == Some(r) && f.postcondition_once((t,), r),
+                })]
+                fn map<U, F>(self, f: F) -> Option<U>
+                where
+                    F: FnOnce(T) -> U;
+
+                #[requires(match self {
+                    None => true,
+                    Some(t) => f.precondition((&t,)),
+                })]
+                #[ensures(result == self)]
+                #[ensures(match self {
+                    None => true,
+                    Some(t) => f.postcondition_once((&t,), ()),
+                })]
+                fn inspect<F>(self, f: F) -> Option<T>
+                where
+                    F: FnOnce(&T);
+
+                #[requires(match self {
+                    None => true,
+                    Some(t) => f.precondition((t,)),
+                })]
+                #[ensures(match self {
+                    None => result == default,
+                    Some(t) => f.postcondition_once((t,), result)
+                })]
```

Comment:
These kinds of specs really make me wish for a predicate-transformer style specification, or some alternate mechanism to reduce this duplication.